packages/components/src/components/permissions/PermissionAssignments.tsx

```diff
@@ -0,0 +1,154 @@
+    typesToShow?: List<string>
+    showDetailsPanel?: boolean
+}
+
+interface State {
+    selectedPrincipal: Principal
+    dirty: boolean
+    submitting: boolean
+    saveErrorMsg: string
+}
+
+export class PermissionAssignments extends React.PureComponent<Props, State> {
+
+    static defaultProps = {
+        title: 'Security roles and assignments',
+        typesToShow: List<string>(['g', 'u']),
+        showDetailsPanel: true
+    };
+
+    constructor(props: Props) {
+        super(props);
+
+        this.state = {
+            selectedPrincipal: undefined,
+            dirty: false,
+            submitting: false,
+            saveErrorMsg: undefined
+        };
+    }
+
+    addAssignment = (principal: Principal, role: SecurityRole) => {
+        const { policy, onChange } = this.props;
+        onChange(SecurityPolicy.addAssignment(policy, principal, role));
+        this.setState(() => ({selectedPrincipal: principal, dirty: true}));
+    };
+
+    removeAssignment = (userId: number, role: SecurityRole) => {
+        const { policy, onChange } = this.props;
+        onChange(SecurityPolicy.removeAssignment(policy, userId, role));
+        this.setState(() => ({selectedPrincipal: undefined, dirty: true}))
+    };
+
+    showDetails = (selectedUserId: number) => {
+        this.setState(() => ({selectedPrincipal: this.props.principalsById.get(selectedUserId)}));
+    };
+
+    onSavePolicy = () => {
+        const { containerPath, policy} = this.props;
+
+        this.setState(() => ({submitting: true}));
+
+        Security.savePolicy({
+            containerPath: containerPath,
+            policy: {policy},
+            success: (response) => {
+                if (response.success) {
+                    this.props.onSuccess();
+                    this.setState(() => ({selectedPrincipal: undefined, submitting: false, dirty: false}));
+                }
+                else {
+                    // TODO when this is used in LKS, need to support response.needsConfirmation
+                    const message = response.message.replace('Are you sure that you want to continue?', '');
+                    this.setState(() => ({saveErrorMsg: message, submitting: false}));
+                }
+            },
+            failure: (response) => {
+                this.setState(() => ({saveErrorMsg: response.exception, submitting: false}));
+            }
+        });
+    };
+
+    render() {
+        const { title, policy, rolesToShow, typesToShow, roles, rolesByUniqueName, principals, error, showDetailsPanel } = this.props;
+        const { selectedPrincipal, saveErrorMsg, submitting, dirty } = this.state;
+        const isLoading = (!policy || !roles || !principals) && !error;
+
+        if (isLoading) {
+            return <LoadingSpinner/>
+        }
+        else if (error) {
+            return <Alert>{error}</Alert>
+        }
+
+        // use the explicit set if passes as a prop, fall back to the relevant roles for the policy
```

Comment:
I'm not quite understanding this comment.  Probably `passes` is meant to be `passed`.  But is this saying "Use the explicit set if passed as a prop and otherwise default to the relevant roles for the policy?"

Reply:
I'll update the comment to fix my typo and clear things up. Yes, the code will use the prop if provided and fall back to the relevant roles in the policy if the prop is undefined.